src/MvvmNano.Forms/MvvmNanoFormsPresenter.cs

```diff
@@ -218,6 +232,30 @@ protected virtual Task OpenPageAsync(Page page)
             return CurrentPage.Navigation.PushAsync(page, true);
         }
 
+        /// <summary> 
+        /// Makes sure that a detail page gets opened as detail page if a MasterDetailPage is set as MainPage.
+        /// This method should be called bevore <see cref="OpenPageAsync"/>. 
+        /// </summary>
+        /// <typeparam name="TViewModel"></typeparam>
+        /// <param name="page"></param>
+        private async Task StartOpeningPage<TViewModel>(Page page)
+        { 
+            if (IsMasterDetailApplication()
+                && ((MvvmNanoMasterDetailApplication)Application).MasterPage != null 
+                && ((MvvmNanoMasterDetailApplication)Application).MasterDetails.FirstOrDefault(o => o.ViewModelType == typeof (TViewModel)) != null) //Check if the new page is a detail page
+            {
+                //Check if the current page is opened as modal page and close it if that is the case.
+                if (((MvvmNanoMasterDetailApplication)Application).MasterPage.Detail.Navigation.ModalStack.Any()
+                    && ((MvvmNanoMasterDetailApplication)Application).MasterPage.Detail.Navigation.ModalStack.FirstOrDefault() == CurrentPage)
+                {
+                    await ((MvvmNanoMasterDetailApplication)Application).MasterPage.Detail.Navigation.PopModalAsync();
+                }
+                ((MvvmNanoMasterDetailApplication)Application).MasterPage.SetDetail(page);
+            } 
+            else
+                await OpenPageAsync(page); 
```

Comment:
I think you overlooked, that `await masterDetailApplication.MasterPage.Detail.Navigation.PopModalAsync();` in line 251 is followed by `masterDetailApplication.MasterPage.SetDetail(page);'